chain/network/src/accounts_data/mod.rs

```diff
@@ -62,24 +70,113 @@ pub struct CacheSnapshot {
+    /// 0, until it learns from the network about data it already signed in the
+    /// previous execution. It means that a node may sign 2 data with the exact same
+    /// version. To avoid an inconsistent state of the network (i.e. situation in which
+    /// some nodes will learn about one data with the given version, some about the other)
+    /// we introduce a tie breaker in a form of UTC timestamp of signing (note that here
+    /// we do not rely on clock monotonicity, which is not guaranteed for UTC clock,
+    /// just assume that timestamp collision is unlikely).
+    ///
+    /// The alternatives to using a timestamp would be:
+    /// * adding a random_minor_version to AccountData, specifically to avoid collisions
+    ///   (so we would be comparing `(version,random_minor_version)` instead)
+    /// * using some crypto hash function `h` and compare `(version,h(data))`. Assuming that `h`
+    ///   behaves like a random oracle, the semantics will be equivaluent to
+    ///   `random_minor_version`, except that if a node signs exactly the same data and in the
+    ///   previous run, then there will be a collision. But in such a case it doesn't matter
+    ///   since the data is the same.
+    /// * storing `version` of the last signed AccountData in persistent storage,
+    ///   so that the node literally never signs AccountsData with colling versions.
+    ///   This assumption is fragile as long as validators migrate their nodes without copying over
+    ///   the whole storage.
     fn is_new(&self, d: &SignedAccountData) -> bool {
         self.keys.contains(&d.account_key)
             && match self.data.get(&d.account_key) {
-                Some(old) if old.version >= d.version => false,
+                Some(old) if (old.version, old.timestamp) >= (d.version, d.timestamp) => false,
                 _ => true,
             }
     }
 
-    fn try_insert(&mut self, d: Arc<SignedAccountData>) -> Option<Arc<SignedAccountData>> {
+    /// Inserts d into self.data, if
+    /// * `d.account_data` is in self.keys AND
+    /// * `d.version > self.data[d.account_data].version`.
+    /// If d would override local for this node, an AccountData based on `self.local` is signed
+    /// and inserted instead to rollback the overriding change (it can happen in case the node has
+    /// been restarted and we observe the old value emitted by the previous run).
+    /// It returns the newly inserted value (or None if nothing changed).
+    /// The returned value should be broadcasted to the network.
+    fn try_insert(
+        &mut self,
+        clock: &time::Clock,
+        d: Arc<SignedAccountData>,
+    ) -> Option<Arc<SignedAccountData>> {
         if !self.is_new(&d) {
             return None;
         }
+        let d = match &self.local {
+            Some(local) if d.account_key == local.signer.public_key() => Arc::new(
+                VersionedAccountData {
```

Comment:
shouldn't we do it only if account data differs?

Reply:
we could, but we don't care. If we encounter anything that we don't know that we have signed, we override it. It is a form of defense in depth - especially since the underlying proto may contain extra information that we cannot interpret (the received data might have been signed by a different neard version).